### ISSUE TITLE
Designing /api/v2

### ISSUE THREAD

**Issue by nvictus** [higlass/higlass#642]
After discussing with Peter on Slack, let's start a discussion about what v2 might look like:

These are the endpoints we have now:
```
/chrom-sizes/
/uids_by_filename/
/tilesets/
/tileset_info/
/tiles/
/register_url/
```

If this were RESTful, how many resources would there be? I think there's just 3:
* `/chrom-sizes/` or `/coord-systems/`
* `/tilesets/`
* `/tiles/`

`/register_url/` ideally should be some kind of POST request to /tilesets/... a way to register a new remote tileset.

`/tileset_info/` would just be a GET request to `/tilesets/` with a id parameter

`/uids_by_filename/` would just be a different GET request to `/tilesets/`

Thoughts?

**Comment by flekschas** [comment]
Thanks for getting this discussion started. Without knowing the prior discussion it would be helpful to point out what the current limitations are to know how we can best address them. RESTfulness is nice to have but certainly not needed. Some things are even harder / less performant with a pure RESTful interface.

In general, `fragments_by_loci ` needs to be there too unless we want to drop support for HiPiler. 
Its API docs are here: http://higlass.io/api/v1/fragments_by_loci/

_Note, that this endpoint is not RESTful and it's hard to make it RESTful as you have to send state in a POST request to retrieve the snippets._

> `/register_url/` ideally should be some kind of POST request to `/tilesets/...` a way to register a new remote tileset.

If a remote tileset and a local tileset behave in the same way then I agree, otherwise I would treat them as two different resources. Generally, a PUT request would be safer though due to idempotency.

> `/uids_by_filename/` would just be a different GET request to `/tilesets/`

How can there be different GET requests to `/tilesets/`? I might be wrong but I think as soon as you send any state with a GET request the API is not RESTful anymore. I am not saying that we need a pure REST API. To decide that we should probably point out which parts of the current API could be improved.

**Comment by nvictus** [comment]
I agree that there's no need to be purely RESTful. But it would still be nice to make it clear what resources exist and consolidate unnecessary methods like `tileset_info`.

> How can there be different GET requests to /tilesets/? 

Sorry, I was being sloppy. I meant that it could be a response to a different set of input arguments. uids_by_filename is basically an alternative way for the server to format the `/tilesets/` GET query, no?

> If a remote tileset and a local tileset behave in the same way then I agree, otherwise I would treat them as two different resources.

Good question. In higlass-python, the remote tileset registration merely locates the url on a httpfs mount and stores that location as a file path. Otherwise it behaves just like a local tileset. Though, I don't think all remote tilesets need to be mounted over httpfs. Some backends should be able to handle remote resources naturally. This is a server-side implementation detail, so I'm not sure whether they should be treated as different resources or not.

> Generally, a PUT request would be safer though due to idempotency.

Yes, good RESTing :)

> In general, fragments_by_loci needs to be there too unless we want to drop support for HiPiler.

Yes, good point! What other methods did I miss in my list above?

> it would be helpful to point out what the current limitations are to know how we can best address them.

The goal would be to document a spec that can be used to implement functional higlass servers, like the one I cobbled together in [higlass-python](https://github.com/higlass/higlass-python/blob/master/higlass/server.py#L52). In higlass-python we want to find the lowest-barrier way to design a custom tileset (and/or coordinate system) and serve it to a higlass client, either by (1) constructing a viewconf programmatically, or by (2) constructing a track source server programmatically.... and (1) is actually achieved through (2), so this is really all about (2).

e.g., the flask server in higlass-python currently doesn't implement `fragments_by_loci`.

**Comment by nvictus** [comment]
Even putting v2 aside, maybe we should have a proper API reference for all v1 endpoints (http://docs.higlass.io/higlass_server.html#api)

**Comment by nvictus** [comment]
Update. I was really only listing the tilesets API, which is the minimal set for a track source server. But here is a more complete listing.

### Tilesets
```
/chrom-sizes/
/uids_by_filename/
/tilesets/
/tileset_info/
/tiles/
/register_url/
```

### Viewconfs
```
/viewconf/
```

### Completion
```
/suggest/
```

### Fragments
```
/fragments_by_loci/
/fragments_by_chr/
/loci/
```

**Comment by pkerpedjiev** [comment]
I'm not aware that we're actually using `uids_by_filename` for anything. It's a good function to have for checking before deleting tilesets. In an ideal world, each tileset should point to its own file. In the less than ideal world that we live, multiple tilesets could point to the same file.

When deleting a file, all tilesets should be checked to make sure they don't point to it.

> which is the minimal set for a track source server

What's your definition of a minimal track source server? For viewing data, only `/tileset_info` and `/tiles` are necessary. For listing available tilesets we add `/tilesets`. For saving viewconfs `/viewconfs`.

There are often cases where I only implement tileset_info and tiles because I use `higlass-python` to view a dataset and I'm not interested in having it listed in the track listings, storing viewconfs or using hipiler.

With that said, I think your classification makes a lot of sense. Most important, IMO, is documentation for the current API. I know flask has some add-ons that auto-document APIs and Django probably does too. Haven't spent enough time looking at them to be familiar.

**Comment by nvictus** [comment]
> What's your definition of a minimal track source server?

I think a proper track source server should provide its tileset listing, so that it can be registered with a viewer-client and display its offerings. `tileset_info` and `tiles` are probably a minimal API for a non-public track source server.

**Comment by alexpreynolds** [comment]
Does use of tilesets?ac=foo do the same as uids_by_filename?

$ curl -qL "http://127.0.0.1/api/v1/tilesets?ac=foo" | jq .results[0].uuid
"Vuol_ez-QlO98FsSgtwLyQ"

I didn't know that tilesets could point to the same filename. I thought
that ingesting a tileset adds a suffix if the filename already exists.

Non-RESTful utility endpoints are helpful. Using filter options with
existing endpoints works, too, but the higlass API documentation might lack
some detail on these (https://docs.higlass.io/higlass_server.html#api). It
is definitely helpful to be able to get UUIDs from filenames for my use
case, but I had to dig into the higlass-server code to find other filter
options.

I don't have much useful to add to the v2 discussion, but thought a
perspective from the user side about endpoints might be helpful.
